### PR TITLE
main/linux-rpi: Build a subpackage for RPi 4

### PR DESCRIPTION
**Pull request by filoozom**
The Raspberry Pi 4 needs to be built with a different configuration: https://www.raspberrypi.org/documentation/linux/kernel/building.md

### CODE REVIEW

File: main/linux-rpi/APKBUILD [hunk 1]
```diff
@@ -24,10 +24,13 @@ source="https://kernel.org/pub/linux/kernel/v4.x/linux-$_kernver.tar.xz
...
 
 	config-changes-rpi.armhf
+	config-changes-rpi.armv7
 	config-changes-rpi.aarch64
 	config-changes-rpi2.armhf
-	config-changes-rpi.armv7
 	config-changes-rpi2.armv7
+	config-changes-rpi4.armhf
+	config-changes-rpi4.armv7
+	config-changes-rpi4.aarch64
```
Comment (clandmeter) [inline main/linux-rpi/APKBUILD]: Where is the package function for rpi4?

Comment (filoozom) [reply main/linux-rpi/APKBUILD]: Looks like that did it, thanks! Although the error message could've been a bit clearer:
```
>>> linux-rpi-dev: Running split function _dev...
make: Entering directory '/aports/main/linux-rpi/src/linux-4.14'
make[1]: Entering directory '/aports/main/linux-rpi/pkg/linux-rpi-dev/usr/src/linux-headers-4.14.78-0-rpi'
/aports/main/linux-rpi/src/linux-4.14/Makefile:509: arch/aarch64/Makefile: No such file or directory
make[2]: *** No rule to make target 'arch/aarch64/Makefile'.  Stop.
make[1]: *** [/aports/main/linux-rpi/src/linux-4.14/Makefile:264: __build_one_by_one] Error 2
make[1]: Leaving directory '/aports/main/linux-rpi/pkg/linux-rpi-dev/usr/src/linux-headers-4.14.78-0-rpi'
make: *** [Makefile:146: sub-make] Error 2
make: Leaving directory '/aports/main/linux-rpi/src/linux-4.14'
>>> ERROR: linux-rpi-dev: _dev failed
>>> ERROR: linux-rpi: prepare_subpackages failed
>>> ERROR: linux-rpi: rootpkg failed
```

**Comment by ncopa** [comment]
Do we really want support 2 different 32bit arm architectures (armv6 and armv7) for rpi4? Wouldn't it be enough with only armv7 + aarch64?

Can you please also squash your commits and fix the commit message?
See https://github.com/alpinelinux/aports/blob/master/.github/CONTRIBUTING.md#creating-a-pull-request-pr

**Comment by filoozom** [comment]
@ncopa Very valid point actually. I guess we could scratch `armhf`. Maybe even `armv7`? Although having a 32 bit option could still be interesting for some people I guess.

I was going to squash the commits once everything works correctly, because I find it easier to review single commits than the whole thing every time. I can do it now if you want though.

**Comment by clandmeter** [comment]
> @ncopa Very valid point actually. I guess we could scratch `armhf`. Maybe even `armv7`? Although having a 32 bit option could still be interesting for some people I guess.
> 
> I was going to squash the commits once everything works correctly, because I find it easier to review single commits than the whole thing every time. I can do it now if you want though.

Please keep in mind rpi project does not officially support 64bit mode. raspbian is still only available with 32bit kernel (afaik).
I dont mind dropping armhf but we should probably keep armv7.

**Comment by Cogitri** [comment]
Although it's pretty PR-y there's https://www.balena.io/blog/balena-releases-first-fully-functional-64-bit-os-for-the-raspberry-pi-4/ now which kind of is an official 64-bit OS for the RPi4.
I'd definitely vote for keeping armv7 too, I don't have a strong opinion about armhf

**Comment by lygstate** [comment]
Ping for RPi4:) I also wanna running alpine on RPi 4

**Comment by ncopa** [comment]
Lets keep armv7 and aarch64 unless upstream explicitly says that 64 bit is unsupported.

**Comment by filoozom** [comment]
> Lets keep armv7 and aarch64 unless upstream explicitly says that 64 bit is unsupported.

Well, `aarch64` was already supported for RPi 3 on Alpine, so it would be strange to remove it, right?

**Comment by CvH** [comment]
was anyone able to build this ? if so could someone at least link the kernel.apk ?
I tried now for days (building is horrible slow) but all I got was an error at the end of the actually build job after the kernel was build successfully 
```
entering fakeroot
./abuild can't cd /path/aports/main/linux-rpi/... error folder not existing
```
Folder and path exist ofc and is accessible for the build user/group.
Strangely this also kills off the screen session where it was run. 

I run different stuff over the weekend but always got the same error (not too sure what the last approach gave me, as screen session was gone).
Maybe this is wrong idk.
```
cd aports/main/linux-rpi/
abuild -crK
```

**Comment by PenguinOwl** [comment]
Just a +1 for a quick and speedy merge.

**Comment by filoozom** [comment]
@CvH What machine are you building on?

**Comment by CvH** [comment]
@filoozom tried it at Qemu (which is HORRIBLE slow), then I tired it at an RPi3 and it is still ultra slow and it error out with the same problem

Just to bring some light into that, there is no doc existent that shows how to build kernel into an image or how the whole stuff actually works. Google is no help here because there is no proper source for informations like a forum or a well used mailing list. Most stuff is done at IRC where every gathered information goes /dev/null immediately due not exiting logs. So any gained knowledge is directly lost.

Informations I gathered, could be wrong, nobody knows - you can't get any of it from an official source

**The official alpine way is: **
- Rent/buy an beefy arm server, compile it there.
- Qemu is there but nobody uses it because it is not usable slow and the created bins are not reliable - thats why "nobody" uses it anywhere for this task.
- There is no cross compiling (even there is some scripts).
- Creating release images with changed kernels/uboot is basically not possible due bugs - or not understanding how it works (no docs no fun).
- The development happens at https://gitlab.alpinelinux.org/alpine - github looks like some additional source that is only visited once in a month, because everyone loves to open accounts at 3rd party github clones.
- There are no proper up to date docs, if you want something go to IRC.


I basically gave up at the moment, I have no time for debugging stuff that needs 12h to compile. I have other (all ARM related) changes and fixes that I can't even test because I can't build it.

**Comment by lygstate** [comment]
I would suggest merge linux-rpi into linux-vanilla for reduce the maintain burden, refer to https://gitlab.alpinelinux.org/alpine/aports/issues/10974

**Comment by CvH** [comment]
@lygstate you still need the RPi Kernel, current mainline is currently not good enough and missing a lot fixes
this will change in future but not anytime soon

**Pull request closed without merging**